extensions/amp-brightcove/1.0/component.js

```diff
@@ -0,0 +1,225 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '#preact';
+import {addParamsToUrl} from '../../../src/url';
+import {useCallback, useMemo, useRef, useState} from '#preact';
+import {forwardRef} from '#preact/compat';
+import {VideoIframe} from '../../amp-video/1.0/video-iframe';
+import {mutedOrUnmutedEvent} from '../../../src/iframe-video';
+import {VideoEvents} from '../../../src/video-interface';
+import {dispatchCustomEvent} from '#core/dom';
+
+/** @const {string} */
+const DEFAULT = 'default';
+
+/**
+ * @enum {string}
+ * @private
+ */
+const methods = {
+  'mute': 'muted',
+  'unmute': 'muted',
+  'showControls': 'controls',
+  'hideControls': 'controls',
+};
+
+/**
+ * @enum {string}
+ * @private
+ */
+const events = {
+  'ready': VideoEvents.LOAD,
+  'playing': VideoEvents.PLAYING,
+  'pause': VideoEvents.PAUSE,
+  'ended': VideoEvents.ENDED,
+  'ads-ad-started': VideoEvents.AD_START,
+  'ads-ad-ended': VideoEvents.AD_END,
+  'loadedmetadata': VideoEvents.LOADEDMETADATA,
+};
+
+/**
+ * @param {!BrightcoveDef.Props} props
+ * @param {{current: ?VideoWrapperDef.Api.CarouselApi}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function BrightcoveWithRef(
+  {
+    account,
+    autoplay,
+    embed = DEFAULT,
+    muted: mutedProp,
+    onReady,
+    player = DEFAULT,
+    playlistId,
+    referrer,
+    videoId,
+    onPlayingState,
+    urlParams,
+    ...rest
+  },
+  ref
+) {
+  const playerStateRef = useRef({});
+  const [muted, setMuted] = useState(mutedProp);
+  const src = useMemo(
+    () =>
+      addParamsToUrl(
+        `https://players.brightcove.net/${encodeURIComponent(account)}` +
+          `/${encodeURIComponent(player)}` +
+          `_${encodeURIComponent(embed)}/index.html` +
+          '?amp=1' +
+          // These are encodeURIComponent'd in encodeId_().
+          (playlistId
+            ? '&playlistId=' + encodeId(playlistId)
+            : videoId
+            ? '&videoId=' + encodeId(videoId)
+            : ''),
+        {...urlParams, referrer, playsinline: true, autoplay: undefined}
+      ),
+    [account, embed, player, playlistId, referrer, urlParams, videoId]
+  );
+
+  const makeMethodMessage = useCallback((method) => makeMessage(method), []);
```

Comment:
I don't think we need `useCallback` here. The original method which is already hoisted outside the component definition will not change identity (and you aren't modifying it here)

Reply:
Agreed, we can just pass the higher-order function directly.